### DISCUSSION TITLE
Issue on connecting multiple UEs in docker environment (ZMQ mode).

### DISCUSSION THREAD

**Discussion by Ali78k** [srsran/srsRAN_Project#499] [General Help]
I've emulated successfully multiple UEs scenario by following [this instruction](https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#multi-ue-emulation) and provided configurations, but there is a problem by using these configurations in `docker compose` environment.
I've defined a docker network for simulated air interface, and some IPs in the subnet are set to gNB and three UEs. Nevertheless, when the 5GC, gNB, and UEs containers are up, I run the provided grc file -that IP addresses are changed correspondingly- I face the following error:
```
Warning: failed to XInitThreads()
Traceback (most recent call last):
  File "/root/testbedflexricnew/multi_ue_scenario.py", line 227, in <module>
    main()
  File "/root/testbedflexricnew/multi_ue_scenario.py", line 205, in main
    tb = top_block_cls()
  File "/root/testbedflexricnew/multi_ue_scenario.py", line 100, in __init__
    self.zeromq_req_source_1_0 = zeromq.req_source(gr.sizeof_gr_complex, 1, 'tcp://172.22.0.44:2201', zmq_timeout, False, zmq_hwm)
RuntimeError: Invalid argument

```
How can I fix it?
Any hint will be appreciated.

**Comment by pgawlowicz** [comment]
Hi,
We do not provide a docker compose file in the tutorial, therefore we do not really know how your setup looks like. 
It seems that gnuradio has some issues when opening a ZMQ socket. Are you sure this is a right IP address? also did you expose port 2201 in the docker container?
Best,
Piotr

**Comment by Ali78k** [comment]
Dear @pgawlowicz thanks for your reply. Sorry for lack of sufficient information.
I've written three UE services in docker compose file along with other services (gNB, 5GC, etc), which each one includes some ENV variables that define TX and RX IP:PORTs. IPs are set in `networks` subsection and ports are defined in `expose` subsection. Probably the problem is derived from TX and RX IP and ports.
What should these IP:PORTs be defined precisely?
Moreover, I have seen in a forked project (OAIC at [this link](https://openaicellular.github.io/oaic/MultipleUEs.html#start-a-multi-ue-network)), that the `*` was used in introducing TX IP (i.e. `tcp:\\*:port`). The reason of the usage is not clear for me.

Sincerely,
Thanks for your time.

**Comment by Ali78k** [comment]
In the gNB/UE configuration, I believe that the `tx_port` corresponds to the gNB/UE IP, while the `rx_port` is associated with the UE/gNB IP, as discussed in [this thread](https://github.com/srsran/srsRAN_Project/discussions/406).
I have documented the details of my Docker environment in an image. it is clear to define all Tx/Rx parameters except gNB Rx. My assumption is that specifying port mapping through the 'ports' subsection in Docker Compose would resolve the issue. By performing this action, I can configure gNB Rx to the IP address 172.22.0.1, along with the specified port (i.e. 2001).
![image](https://github.com/srsran/srsRAN_Project/assets/101015446/883040b6-9987-4ab4-8f3b-5ca598997a7c)

However, is there a possibility of defining multiple `rx_ports` in the gNB configuration?










Thanks in advance.

**Reply by pgawlowicz** [reply]
Hi, You cannot connect multiple UEs directly to the gNB. Instead, you need to use a signal broker in between. Please have a closer look at this tutorial section: https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#multi-ue-emulation

You can compare the Gnuradio flowgraph and the IP/Ports in ZMQ blocks with the RF device section in UE and gnb configs to find out how the wireless channel connected.

**Reply by pgawlowicz** [reply]
@Ali78k any update on this issue?

**Reply by Ali78k** [reply]
Thanks for your assistance, @pgawlowicz . I apologize for the delay in getting back to you. Despite numerous attempts, I've been unable to uncover any viable workarounds. Finally, I've decided to run the UEs normally, without utilizing Docker.
